mapping_parenting_tech/analysis/prototyping/play_store/fn_tester.py

```diff
@@ -0,0 +1,140 @@
+#     display_name: Python 3.8.0 ('mapping_parenting_tech')
+#     language: python
+#     name: python3
+# ---
+
+# %% [markdown]
+# # Testing notebook
+# Testing refactoring of functions to manage ids, details and reviews of Play Store apps
+
+# %%
+from mapping_parenting_tech.utils import play_store_utils as psu
+from mapping_parenting_tech import PROJECT_DIR, logging
+
+# %% [markdown]
+# ## Working with app ids
+# ### Getting app ids from source/s
+# App ids are retrieved in two ways:
+# 1. parsing an HTML file saved from a Play Store web page - this can be done with a single web page or via a folder that contains multiple web pages
+# 2. snowballing from a given app id and using that to identify related apps
+
+# %%
+# test downloading apps from a single HTML web page
+appIds = psu.get_playstore_app_ids_from_html("play_store/parenting_top_free.html")
+appIds
+
+# %%
+# test parsing a folder of HTML files
+appIds = psu.get_playstore_app_ids_from_folder("play_store/education_apps")
+len(appIds)
+
+# %%
+# test retrieving related apps to a depth of 3
+appIds = psu.app_snowball("uk.org.bestbeginnings.babybuddymobile", 3)
+len(appIds)
+
+# %% [markdown]
+# ### Saving app ids and loading saved app ids
+# App ids are saved to a single CSV file `outputs/data/all_app_ids.csv` with one app id per line
+
+# %%
+# test saving new apps - dry run allows us to do this without updating the file itself
+# function returns complete list of new apps (i.e., not just those that have been added)
+new_appIds = psu.update_all_app_id_list(appIds, dry_run=False)
+new_appIds
+
+# %%
+# load all app ids
+all_appIds = psu.load_all_app_ids()
+len(all_appIds)
+
+# %%
+# check if an app is in the list of all app ids - here we're using the results of the snowball above
+compare_list = psu.is_app_in_list(appIds)
+
+# enumerate through compare_list to identify apps that aren't present
+print(f"{compare_list.count(False)} apps are not already saved")
+for i, present in enumerate(compare_list):
+    if not present:
+        print(appIds[i])
+
+# %% [markdown]
+# ## Working with app details
+# App details can be downloaded into memory in a standalone fashion or saved to and loaded from disk via the JSON file: `outputs/data/all_app_details.json`. This file is updated/synchronised according to the saved app ids.
+
+# %%
+# get the app details for the list of apps in `app_list`
+
+app_list = [
+    "com.amboss.medical.knowledge",
+    "com.excelatlife.depression",
+    "au.com.sightwords.parrotfish.lite",
+    "com.kitefaster.whitenoise",
+    "com.apololo.brightest.flashlight",
+]
+app_list_details = psu.get_playstore_app_details(app_list)
+app_list_details["com.amboss.medical.knowledge"]["title"]
+
+# %%
+# load all app details into a dict
+all_app_details = psu.load_all_app_details()
+len(all_app_details)
+
+# %%
+# update the details for apps whose ids are already saved - this returns the details for all of the apps in question
+all_app_details = psu.update_all_app_details()
```

Comment:
Not exactly sure how to interpret this - looks like same apps were not found. Is that expected?
```
2022-02-25 17:40:29,257 - root - INFO - There are 8 apps whose details will be updated.
Retrieving app details:   0%|                                                                                                                            | 0/8 [00:00<?, ?it/s]
2022-02-25 17:40:29,377 - root - WARNING - Error on app id com.kids.free.audiobook.lullaby.songs.for.baby.offline: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  12%|██████████████▌                                                                                                     | 1/8 [00:00<00:00,  8.29it/s]
2022-02-25 17:40:29,550 - root - WARNING - Error on app id com.infinitylabbdr.zonxpressv5: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  25%|█████████████████████████████                                                                                       | 2/8 [00:00<00:00,  6.59it/s]
2022-02-25 17:40:29,673 - root - WARNING - Error on app id com.cocomelonallsongs21.furiki: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  38%|███████████████████████████████████████████▌                                                                        | 3/8 [00:00<00:00,  7.23it/s]
2022-02-25 17:40:29,785 - root - WARNING - Error on app id com.tocabocaguide.LifeAppkaloco: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  50%|██████████████████████████████████████████████████████████                                                          | 4/8 [00:00<00:00,  7.73it/s]
2022-02-25 17:40:29,922 - root - WARNING - Error on app id com.katsu.subbeddubbed.anime.rer: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  62%|████████████████████████████████████████████████████████████████████████▌                                           | 5/8 [00:00<00:00,  7.63it/s]
2022-02-25 17:40:30,032 - root - WARNING - Error on app id family.locator.mobile.phone.tracker: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  75%|███████████████████████████████████████████████████████████████████████████████████████                             | 6/8 [00:00<00:00,  8.06it/s]
2022-02-25 17:40:30,151 - root - WARNING - Error on app id zd.power.cash.bvv: App not found(404). NotFoundError('App not found(404).')
Retrieving app details:  88%|█████████████████████████████████████████████████████████████████████████████████████████████████████▌              | 7/8 [00:00<00:00,  8.19it/s]
2022-02-25 17:40:30,268 - root - WARNING - Error on app id ze.x.earn.dve: App not found(404). NotFoundError('App not found(404).')
Retrieving app details: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:01<00:00,  7.90it/s]
2022-02-25 17:40:31,129 - root - INFO - Updated app details file.
```

Reply:
Yes! The Play Store is a dynamic place and apps get taken down/removed. Could add a separate function to check is apps are live, but think this is not urgent.